File: frontend/src/api/capital.ts
/** Capital management API client. */

import client from "./client"

export interface CapitalTransaction {
  id: string
  type: string
  amount: number
  balance_after: number
  trade_id: string | null
  symbol: string | null
  description: string
  created_at: string
}

export interface PositionCapital {
  symbol: string
  stock_name: string
  shares: number
  cost_price: number
  market_value: number
  cost_basis: number
}

export interface CapitalBreakdown {
  available_cash: number
  position_value: number
  total_assets: number
  utilization_rate: number
  positions: PositionCapital[]
  has_initial_deposit: boolean
}

export interface CapitalHistoryResponse {
  transactions: CapitalTransaction[]
  total: number
}

/** Get full capital breakdown (cash + positions). */
export async function getCapitalBalance(): Promise<CapitalBreakdown> {
  const { data } = await client.get<CapitalBreakdown>("/capital/balance")
  return data
}

/** Deposit funds. */
export async function deposit(
  amount: number,
  description?: string,
): Promise<CapitalTransaction> {
  const { data } = await client.post<CapitalTransaction>("/capital/deposit", {
    amount,
    description: description ?? "",
  })
  return data
}

/** Withdraw funds. */
export async function withdraw(
  amount: number,
  description?: string,
): Promise<CapitalTransaction> {
  const { data } = await client.post<CapitalTransaction>("/capital/withdraw", {
    amount,
    description: description ?? "",
  })
  return data
}

/** Get capital transaction history. */
export async function getCapitalHistory(
  limit = 50,
  offset = 0,
  txType?: string,
): Promise<CapitalHistoryResponse> {
  const params: Record<string, string | number> = { limit, offset }
  if (txType) params.tx_type = txType
  const { data } = await client.get<CapitalHistoryResponse>(
    "/capital/history",
    { params },
  )
  return data
}

File: src/web/schemas/capital.py
"""Pydantic models for the capital management system.

Covers transaction records, balance queries, and breakdown views.
"""

from __future__ import annotations

from pydantic import BaseModel, Field


class CapitalTransaction(BaseModel):
    """A single capital transaction (deposit, withdrawal, or trade settlement)."""

    id: str
    type: str  # initial_deposit | deposit | withdrawal | trade_buy | trade_sell | position_liquidation
    amount: float  # signed: + inflow, - outflow
    balance_after: float
    trade_id: str | None = None
    symbol: str | None = None
    description: str = ""
    created_at: str


class CapitalBalance(BaseModel):
    """Current capital balance snapshot."""

    available_cash: float = 0.0
    total_transactions: int = 0
    has_initial_deposit: bool = False


class PositionCapital(BaseModel):
    """Capital tied up in a single position."""

    symbol: str
    stock_name: str
    shares: int
    cost_price: float
    market_value: float
    cost_basis: float  # shares * cost_price


class CapitalBreakdown(BaseModel):
    """Full capital breakdown: cash + positions."""

    available_cash: float = 0.0
    position_value: float = 0.0
    total_assets: float = 0.0
    utilization_rate: float = 0.0  # position_value / total_assets
    positions: list[PositionCapital] = Field(default_factory=list)
    has_initial_deposit: bool = False


class DepositRequest(BaseModel):
    """Request to deposit funds."""

    amount: float = Field(gt=0)
    description: str = ""


class WithdrawRequest(BaseModel):
    """Request to withdraw funds."""

    amount: float = Field(gt=0)
    description: str = ""


class PositionLiquidationRequest(BaseModel):
    """Request to liquidate (clear) a position and recover capital."""

    symbol: str
    stock_name: str
    shares: int = Field(gt=0)
    price: float = Field(gt=0)
    position_id: str


class CapitalHistoryResponse(BaseModel):
    """Paginated capital transaction history."""

    transactions: list[CapitalTransaction] = Field(default_factory=list)
    total: int = 0
